plugins/catalog-backend-module-github/src/providers/GithubEntityProvider.ts

```diff
@@ -0,0 +1,313 @@
+  /**
+   * A Scheduled Task Runner
+   *
+   * {@link @backstage/backend-tasks#PluginTaskScheduler.createScheduledTaskRunner}
+   * to enable automatic scheduling of tasks.
+   */
+  schedule: TaskRunner;
+
+  /**
+   * The logger to use.
+   */
+  logger: Logger;
+
+  /**
+   * Optionally supply a custom credentials provider, replacing the default one.
+   */
+  githubCredentialsProvider?: GithubCredentialsProvider;
+}
+
+type CreateLocationSpec = {
+  url: string;
+  branch: string;
+  file: string;
+};
+
+/**
+ * Provider which discovers catalog files (any name) within a Github Organization.
+ *
+ * Use `GithubEntityProvider.fromConfig(...)` to create instances.
+ *
+ * @public
+ */
+export class GithubEntityProvider implements EntityProvider {
+  private readonly scheduleFn: () => Promise<void>;
+  private connection?: EntityProviderConnection;
+  private readonly githubCredentialsProvider: GithubCredentialsProvider;
+
+  static fromConfig(config: Config, options: GithubEntityProviderOptions) {
+    const integrations = ScmIntegrations.fromConfig(config);
+    if (!integrations) {
+      throw new Error('No integrations found for github');
+    }
+    return new GithubEntityProvider(integrations, options);
+  }
+
+  private constructor(
+    private integrations: ScmIntegrations,
+    private options: {
+      id: string;
+      orgUrl: string;
+      files: Array<string>;
+      logger: Logger;
+      schedule: TaskRunner;
+      githubCredentialsProvider?: any;
+    },
+  ) {
+    options.logger.child({ target: this.getProviderName() });
+
+    this.githubCredentialsProvider =
+      options.githubCredentialsProvider ||
+      DefaultGithubCredentialsProvider.fromIntegrations(integrations);
+
+    this.scheduleFn = this.createScheduleFn(options.schedule);
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.getProviderName} */
+  getProviderName(): string {
+    return `github-entity-provider:${this.options.id}`;
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    return await this.scheduleFn();
+  }
+
+  private createScheduleFn(schedule: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.options.logger.child({
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const gitHubConfig = this.integrations.github.byUrl(
+      this.options.orgUrl,
+    )?.config;
+
+    if (!gitHubConfig) {
```

Comment:
I think this check is the one you wanted in the constructor; it's probably best to explode immediately and not start up, than to discover this way later in the logs.

Come to think of it: The integration instance returned by byUrl is probably a great parameter to the constructor, instead of the whole integrations object. So to clarify:

```ts
fromConfig(...) {
  const integration = integrations.byUrl(...);
  if (!integration) { throw }
  return new GithubEntityProvider(integration, ...);
}
```

this probably makes testing easier as well, since the class can be created with a simpler / more low level mock